### ISSUE TITLE
VSCode / VS Code / Virtual Studio Code Embedding

### ISSUE THREAD

**Issue by clinejj** [warpdotdev/Warp#257]
I largely use the VSCode integrated terminal when developing, due to the easy integration into my workflow and ability to lay out within the VSCode UI.

It'd be great to use Warp, either as the terminal for the integrated terminal prompt (using a custom profile, see https://code.visualstudio.com/docs/editor/integrated-terminal#_terminal-profiles) or even as a separate extension.

**Comment by elviskahoro** [comment]
if I have Warp selected in vscode as my external terminal program, normally I should be able to run the command "Open New External Terminal" and have a new prompt in that program, at the root of the project that I ran the command from. Currently this doesn't work. Focus is switched to Warp but I do not get the prompt in my project folder.

**Comment by elviskahoro** [comment]
@o1lo01ol1o
From #605

### Describe the solution you'd like?

```shell
I'd like to use warp as a terminal in emacs and vscode
```


### Is your feature request related to a problem? Please describe.

```shell
Half my terminal time is spent running cli commands in IDE windows.  I'm aware that I could just switch to a dedicated terminal, but that's a context switch that I nearly never undertake.
```


### Additional context

_No response_

**Comment by zanderwar** [comment]
I need this :) I spend all day in my IDE, using the integrated terminal will be just far more efficient for me until Warp is capable of it

**Comment by NotSoSuper** [comment]
Would love VS Code remote SSH integration with warp, would fully switch over then!

**Comment by paurakhsharma** [comment]
Only thing preventing me from using Warp throughout my day.

**Comment by adigupta13** [comment]
Please add this feature

**Comment by ahinkle** [comment]
Has anyone found a hack or workaround for now? Loving Warp but hate having to jump between the two applications. 🙂

**Comment by zanderwar** [comment]
VS Code and PHPStorm alike only allow shell applications, warp is not a shell application in itself, it wraps itself around one, warp cannot be integrated into these IDE's without sacrificing majority of the features that make Warp a great product, it'll dilute down to just being another Git Bash. Perhaps a plugin is the only way forward here but warp definitely wont be able to integrate into these IDE's by the standard means, like you would with git bash for example.

**Comment by warpdotdevbot** [comment]
Chances are this kind of integration wouldn't technically be possible until a WASM version of Warp exists.

--- original message ---
On April 14, 2022, 10:38 PM EDT ***@***.*** wrote:



VS Code and PHPStorm alike only allow shell applications, warp is not a shell application in itself, it wraps itself around one, warp cannot be integrated into these IDE's without sacrificing majority of the features that make Warp a great product, it'll dilute down to just being another Git Bash. Perhaps a plugin is the only way forward here but warp definitely wont be able to integrate into these IDE's by the standard means, like you would with git bash for example.

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

-- 
You received this message because you are subscribed to the Google Groups "github" group.
To unsubscribe from this group and stop receiving emails from it, send an email to ***@***.***
--- end of original message ---

**Comment by Milo123459** [comment]
https://saiankit30.medium.com/how-to-change-the-integrated-terminal-in-vs-code-from-default-to-pro-iterm-ish-5c958e13aada Could this article help?

**Comment by bmitc** [comment]
> Chances are this kind of integration wouldn't technically be possible until a WASM version of Warp exists.

Is this the main reason WASM is being prioritized over Linux and Windows (in that order)? Is this feature request going into the timeline or being tracked?

I would wager that the amount of overlap between Visual Studio Code (VS Code) users and people who are using or want to use Warp is very high, which includes myself.

**Comment by mnhattt** [comment]
add this feature, pls 👍

**Comment by tanikondadhanram** [comment]
I am really looking forward to this feature.

**Comment by Indicaza** [comment]
> if I have Warp selected in vscode as my external terminal program, normally I should be able to run the command "Open New External Terminal" and have a new prompt in that program, at the root of the project that I ran the command from. Currently this doesn't work. Focus is switched to Warp but I do not get the prompt in my project folder.

Same issue for us IntelliJ guys.  Or you can choose what terminal executable through IDE but warp is different.

**Comment by elviskahoro** [comment]
@bmitc It isn't the primary reason, the nice thing about a WASM version is that users across any platform will be able to use Warp to some degree (in the browser). It's also because our business model (see our [pricing page](https://www.warp.dev/pricing)) is a freemium model. Free for everyone with compelling features that teams and enterprises would want to pay for. A WASM version is both compelling and in many ways a pre-requisite to some of the ideas we have for team and enterprise.

This is my (personal) understanding of the situation, I hope this helps.

**Comment by vjindal0112** [comment]
As I understand, the main use case is switching quickly between the terminal and VSCode. The Hotkey global window might get it 90% of the way there. [Here are the docs](https://docs.warp.dev/features/hotkey-window).

Using it would look something like this: 
https://user-images.githubusercontent.com/14248349/177417141-aefb27a8-cc1e-456c-8d81-0a1f7e04ebcd.mov

Steps
1. Make a new desktop
2. Fullscreen VSCode in the new desktop
3. Set up Hotkey window in Warp
4. Code away

Let me know what you all think!

**Comment by bmitc** [comment]
Thanks for the information @elviskahoro! That was helpful. Looking forward to when I can use Warp on Windows and/or VS Code! I use it on macOS for work and am liking it.

**Comment by Nauxscript** [comment]
> As I understand, the main use case is switching quickly between the terminal and VSCode. The Hotkey global window might get it 90% of the way there. [Here are the docs](https://docs.warp.dev/features/hotkey-window).
> 
> Using it would look something like this: https://www.loom.com/share/5e9e80960f7e45e38316f8cae3416352
> 
> Steps
> 
> 1. Make a new desktop
> 2. Fullscreen VSCode in the new desktop
> 3. Set up Hotkey window in Warp
> 4. Code away
> 
> Let me know what you all think!

its helpful but not enough, for example, I got two or more window of vscode , its the better if I trigger warp global hotkey and activate warp (if warp have not launch will new one) , and then go to the root path of the workspace where I trigger hotkey (maybe I need a vscode hotkey to activate warp and pass the workspace path), but its seem like no way to do like one of those. I find a way to do like this in vscode: `shift` + `command` + `c` can open an external terminal under the current directory, the external terminal should be the default terminal,so in other way, if warp can be set as the mac default terminal, it can work like that. as far as I know, iterm2 can do it.

**Comment by elviskahoro** [comment]
@Nauxscript the main thing is changing the external terminal setting to Warp.app
does this screenshot help
<img width="663" alt="Screen Shot 2022-07-07 at 11 03 55" src="https://user-images.githubusercontent.com/29553206/177806934-8a854b17-be5b-4830-ba21-b34545864def.png">
?

**Comment by dohooo** [comment]
> > its helpful but not enough, for example, I got two or more window of vscode , its the better if I trigger warp global hotkey and activate warp (if warp have not launch will new one) , and then go to the root path of the workspace where I trigger hotkey (maybe I need a vscode hotkey to activate warp and pass the workspace path), but its seem like no way to do like one of those. I find a way to do like this in vscode: `shift` + `command` + `c` can open an external terminal under the current directory, the external terminal should be the default terminal,so in other way, if warp can be set as the mac default terminal, it can work like that. as far as I know, iterm2 can do it.
> 
> Like @elviskahoro said, we recently launched support for `Shift-Cmd-C` from VSCode in the latest release! Check out [the docs here](https://docs.warp.dev/features/integrations-and-plugins). Let me know if that's what you meant and if that helps get us closer to a 90% solution.

Lame solution 😵‍💫, but I'm glad we can finally use it together. 🥳🥳🥳 I love wrap!

**Comment by Nauxscript** [comment]
> > its helpful but not enough, for example, I got two or more window of vscode , its the better if I trigger warp global hotkey and activate warp (if warp have not launch will new one) , and then go to the root path of the workspace where I trigger hotkey (maybe I need a vscode hotkey to activate warp and pass the workspace path), but its seem like no way to do like one of those. I find a way to do like this in vscode: `shift` + `command` + `c` can open an external terminal under the current directory, the external terminal should be the default terminal,so in other way, if warp can be set as the mac default terminal, it can work like that. as far as I know, iterm2 can do it.
> 
> Like @elviskahoro said, we recently launched support for `Shift-Cmd-C` from VSCode in the latest release! Check out [the docs here](https://docs.warp.dev/features/integrations-and-plugins). Let me know if that's what you meant and if that helps get us closer to a 90% solution.

yes i have use `shift` + `cmd` + `c` hotkey a few day. in the last version of vscode (1.68.x), if i use this hotkey to open Warp, the new tab of warp would not open on the vscode workspace root path. But now the tab path is correct in vscode version 1.69.0.

there is the setting for someone want to use Warp on vscode（its not really "on vscode" exactly, but use them in easy and comfortable way）：

```vscode setting.json
```

**Comment by just-jeb** [comment]
@vjindal0112, both solutions (hotkey window and `shift`+`cmd`+`c`) come to solve the same problem but I couldn't make it work together. I.e. when I press `shift`+`cmd`+`c` it opens a new Warp session but in a separate window but when I use the hotkey it opens the pinned window unrelated to the window opened with `shift`+`cmd`+`c`. Also, the hotkey doesn't work if there is no running Warp session.  
Any chance we can make it work together? It would be a sufficient replacement for the VSCode integration.

**Comment by warpdotdev-dx** [comment]
This is really really difficult eng wise. Just being transparent so people don't get their hopes up that this would come out anytime soon. (elvis) Will share more info if things change!

--- original message ---
On April 11, 2022, 10:08 AM EDT ***@***.*** wrote:



Please add this feature

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

-- 
You received this message because you are subscribed to the Google Groups "github" group.
To unsubscribe from this group and stop receiving emails from it, send an email to ***@***.***
--- end of original message ---

**Comment by dannyneira** [comment]
user on #2305 commented:
> I mostly use a terminal embedded in my IDE of choice (PyCharm / IntelliJ). This uses a configurable shell and automatically provides the context i.e. Conda environment and working directory. It would be nice to use warp in this context.

**Comment by wangmir** [comment]
> > its helpful but not enough, for example, I got two or more window of vscode , its the better if I trigger warp global hotkey and activate warp (if warp have not launch will new one) , and then go to the root path of the workspace where I trigger hotkey (maybe I need a vscode hotkey to activate warp and pass the workspace path), but its seem like no way to do like one of those. I find a way to do like this in vscode: `shift` + `command` + `c` can open an external terminal under the current directory, the external terminal should be the default terminal,so in other way, if warp can be set as the mac default terminal, it can work like that. as far as I know, iterm2 can do it.
> 
> Like @elviskahoro said, we recently launched support for `Shift-Cmd-C` from VSCode in the latest release! Check out [the docs here](https://docs.warp.dev/features/integrations-and-plugins). Let me know if that's what you meant and if that helps get us closer to a 90% solution.

@vjindal0112  Can this be integrated with hot key support too? when I open terminal using cmd+shift+c, then it's opening on separated tab. And when using hot key, it opens on the same tab as i use, but in that case the instance is always same.

**Comment by dannyneira** [comment]
@wangmir Take a look at the Global hotkey window feature, it should help with your use case. https://docs.warp.dev/features/windows/hotkey-window

**Comment by flight505** [comment]
+1 This should be the "most wanted feature"!!! Warp's 10.000 USD reward for good old Billy "The integrated" Wizkiss. 

Just imagine the sheer number of Youtube videos that would be made with banging headlines "New AI and terminal in VScode" and so on. Heck, any homeless person with a youtube channel and 30 followers will be putting on Lulumemons and go trawling for likes.. - written by GPT-4.0 (I am not liable)

**Comment by flight505** [comment]
🖖They've gone Plaid! Less than 24 hours from the last feature request and devs releases 2023.03.23 (v0.2023.03.21.08.02) - Absolutely amazing, fantastic product by the way - [Light years from the competition](https://www.youtube.com/watch?v=9mZaKcl6lUE)

* New features: 
    * Added VSCode Insiders as a supported code editor

**By the way, how many Borg does it take to change a light bulb? 💡**
All of them, because they are one! 🤖

**Comment by flight505** [comment]
Why is it that the same three people have been downvoting this feature request? 
- is there some downside to this feature?

**Comment by Milo123459** [comment]
We're downvoting it because you're spamming this issue with comments. People are subscribed to know when this is released, not about people saying +1.

**Comment by omarluq** [comment]
> This is really really difficult eng wise. Just being transparent so people don't get their hopes up that this would come out anytime soon. (elvis) Will share more info if things change!
> […](#)
> --- original message --- On April 11, 2022, 10:08 AM EDT ***@***.*** wrote: Please add this feature — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
> -- You received this message because you are subscribed to the Google Groups "github" group. To unsubscribe from this group and stop receiving emails from it, send an email to ***@***.*** --- end of original message ---

That's true in the case of actually embedding Warp as VsCode's (or any other IDE for that matter) integrated terminal but on the other hand, creating an extension and using WASM and typescript should address a lot of the blockers no?

On a separate note, I've been using Warp for about 2 years now and I love it! it truly elevates my day to day development experience

**Comment by eliasdefaria** [comment]
Echoing my support for this feature. Would be a huge value add to have it baked into my flow with VSCode. This honestly limits my use cases of warp from daily development down to "hey what was that command to kill port x? Oh let me ask warp AI"

**Comment by iantimmis** [comment]
support support

**Comment by lucianthorr** [comment]
I don't know if this is possible but it could be an improved compromise between the two current solutions (Global Hotkey and Cmd+Shift+C)
If there was a way to customize a Warp profile for the Cmd+Shift+C solution, similar to how we can with Global Hotkey, then a user can say "Pin to bottom of active screen" and Cmd+Shift+C could give the user a Warp terminal on the same screen as VSCode with the workspace root.

**Comment by hakobpogh** [comment]
One possible solution would be creating a VS Code extension for Warp. As far as I know, Warp is JavaScript-based, so most of the code could be reused to make the VS Code extension.

**Comment by GDanielRG** [comment]
I would pay a subscription to warp if this feature gets added

**Comment by dannyneira** [comment]
Locking to prevent spam, We'll post any updates on this thread regarding VS Code integration/embedding. It'll be possible once Warp is on the Web via WASM (no eta on the release).

**Comment by dannyneira** [comment]
Hey folks, we're still early stages as we continue to iterate on the features and UX, and there is no set ETA on launch, but here is a quick preview of what we have so far. 

https://x.com/i/status/1871231420838134242

**Comment by dannyneira** [comment]
Hey folks, we're unlocking so that folks can add more 👍 's to the [top of the issue / original comment](https://github.com/warpdotdev/Warp/issues/257#issue-1021191974). 

**Please avoid adding spam comments like "+1" and 👍** (we may relock the post if there is too much spam)

We did some super early work on this [feature](https://x.com/warpdotdev/status/1871231420838134242), but there is still a lot more we need to do to get it over the finish line so we appreciate your patience.

**Comment by aymuos15** [comment]
Just interested in how much bloat this adds.

I imagine this to be an insanely heavy extension. In your internal evals, has this been an issue or am I 
completely out of place?

Right now it takes me 1-3s when I do 'code file'. I do not want to increase that :) Very very eagerly waiting for this though!!

**Comment by adamazad** [comment]
@danielkv I think this issue is no longer relevant when most IDEs like Cursor can generate commands into the CLI.

**Comment by IGLADI** [comment]
> @danielkv I think this issue is no longer relevant when most IDEs like Cursor can generate commands into the CLI.

Warp is more than an AI tool, I'm (and many others) still waiting on this feature mainly as a better vscode terminal emulator

**Comment by maddada** [comment]
Still waiting for this feature. Can't use warp until it comes.

**Comment by RayTCosgrove** [comment]
@dannyneira any update on this? Happy new year

**Comment by maddada** [comment]
@dohooo I respect your opinion but I personally disagree. I need the editor to review the code and manage source control visually. The terminal built into VS Code is not as nice as the one that Warp provides but I still use it because it's integrated.

If Warp could be integrated into VS Code they would get a ton of new users. (I'm using Claude Code CLI, Opencode, and Codex CLI a ton everyday on multiple projects. Barely write any code these days. But still I need to review what goes to production and VS Code/Cursor is the best way to do this)

**Comment by Mohit0233** [comment]
Now that Warp is open source, is this something that could realistically happen with community help, or is the VSCode integration side still too difficult?

**Comment by Amr-Samy** [comment]
I need to use it internally inside antigravity without open it externally
